### PR TITLE
[SYCL-MLIR]: Consider aliased store ops in MemorySelector

### PR DESCRIPTION
**Pull request by etiotto**
When the `MemorySelector` analyzes collected memref accesses in a loop, and exclusively read-only accesses should be considered, take into account potential aliased writes in the loop.

### CODE REVIEW

**Comment by etiotto** [comment]
Depends on https://github.com/intel/llvm/pull/9730

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp [hunk 1]
```diff
@@ -11,6 +11,7 @@
 //
 //===----------------------------------------------------------------------===//
 
+#include "mlir/Analysis/AliasAnalysis.h"
```
Comment (whitneywhtsang) [inline polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:14]: can remove.

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp [hunk 2]
```diff
@@ -126,208 +173,163 @@ class MemorySelector {
...
 
-    affine::MemRefAccess access(op);
-    addMemRefAccess(access);
+    affine::MemRefAccess memRefAccess(op);
+
+    auto it = memRefToMemRefAccesses.find(memRefAccess.memref);
+    if (it == memRefToMemRefAccesses.end())
+      memRefToMemRefAccesses[memRefAccess.memref] = {memRefAccess};
+    else
+      memRefToMemRefAccesses[memRefAccess.memref].push_back(memRefAccess);
```
Comment (whitneywhtsang) [inline polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:227]:
```suggestion
      memRefToMemRefAccesses[memRefAccess.memref].push_back(memRefAccess);
```
can you check if this work?

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp [hunk 3]
```diff
@@ -126,208 +173,163 @@ class MemorySelector {
...
+      if (auto stride = getConstIntegerValue(strideVal, solver)) {
+        bool strideIsLargeEnough = stride->sgt(8) || stride->slt(-8);
+        useSharedMemory = hasTemporalReuse(memRefAccess, threadVars) &&
+                          (stride->isZero() || strideIsLargeEnough);
+        // FIXME: getConstIntegerValue doesn't return zero for
+        // ConstantIndexOp.
+      } else if (auto constVal = dyn_cast<arith::ConstantIndexOp>(
+                     strideVal.getDefiningOp());
+                 constVal.value() == 0)
+        useSharedMemory = hasTemporalReuse(memRefAccess, threadVars);
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp]: done.

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp (hunk 1 above)
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp]: done.

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp [hunk 4]
```diff
@@ -107,17 +109,62 @@ void createLocalBarrier(OpBuilder &builder) {
...
+/// Return true if \p op potentially writes the same memory as \p memRefAccess.
+bool mayConflictWithWrite(affine::MemRefAccess memRefAccess, Operation *op,
+                          AliasAnalysis &AA) {
+  if (op == memRefAccess.opInst || isMemoryEffectFree(op))
+    return false;
+
+  // Conservatively assume operations with unknown memory effects may
+  // conflict.
+  if (!isa<MemoryEffectOpInterface>(op) &&
+      !op->hasTrait<OpTrait::HasRecursiveMemoryEffects>())
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:153]: In `mayConflictWithWriteInLoop`, we traverse all operations including operations nested inside operations, so we already handle operation with `HasRecursiveMemoryEffects`.

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp (hunk 2 above)
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp]: done.

**Review by whitneywhtsang — APPROVED** [review]

**Comment by etiotto** [comment]
This is a WIP feature, so merging with one approval. Post-commit review comments are welcome.

**Pull request merged**